### ISSUE TITLE
cannot compile cuda preset on osx

### ISSUE THREAD

**Issue by RoelVanderPaal** [bytedeco/javacpp-presets#57]
Steps taken:

1. Install CUDA 7.0
1. Copy cudnn.h to /usr/local/cuda/include
1. Copy libcudnn.6.5.dylib, libcudnn.dylib, libcudnn_static.a to /usr/local/cuda/lib
1. run 'mvn install' on javacpp-presets/cuda

Error:
```
[INFO] Generating /Users/roel/workspace/javacpp-presets/cuda/target/classes/org/bytedeco/javacpp/jnicudnn.cpp
[INFO] Compiling /Users/roel/workspace/javacpp-presets/cuda/target/classes/org/bytedeco/javacpp/macosx-x86_64/libjnicudnn.dylib
[INFO] g++ -I/usr/local/cuda/include/ -I/Library/Java/JavaVirtualMachines/jdk1.8.0_20.jdk/Contents/Home/include -I/Library/Java/JavaVirtualMachines/jdk1.8.0_20.jdk/Contents/Home/include/darwin /Users/roel/workspace/javacpp-presets/cuda/target/classes/org/bytedeco/javacpp/jnicudnn.cpp -Wl,-rpath,@loader_path/. -march=x86-64 -m64 -Wall -O3 -fPIC -dynamiclib -o /Users/roel/workspace/javacpp-presets/cuda/target/classes/org/bytedeco/javacpp/macosx-x86_64/libjnicudnn.dylib -L/usr/local/cuda/lib/ -Wl,-rpath,/usr/local/cuda/lib/ -lcudnn -lcuda -lcudart -framework JavaVM 
Undefined symbols for architecture x86_64:
  "_cudnnGetPooling2dForwardOutputDim", referenced from:
      _Java_org_bytedeco_javacpp_cudnn_cudnnGetPooling2dForwardOutputDim__Lorg_bytedeco_javacpp_cudnn_00024cudnnPoolingStruct_2Lorg_bytedeco_javacpp_cudnn_00024cudnnTensorStruct_2_3I_3I_3I_3I in jnicudnn-910341.o
      _Java_org_bytedeco_javacpp_cudnn_cudnnGetPooling2dForwardOutputDim__Lorg_bytedeco_javacpp_cudnn_00024cudnnPoolingStruct_2Lorg_bytedeco_javacpp_cudnn_00024cudnnTensorStruct_2Lorg_bytedeco_javacpp_IntPointer_2Lorg_bytedeco_javacpp_IntPointer_2Lorg_bytedeco_javacpp_IntPointer_2Lorg_bytedeco_javacpp_IntPointer_2 in jnicudnn-910341.o
      _Java_org_bytedeco_javacpp_cudnn_cudnnGetPooling2dForwardOutputDim__Lorg_bytedeco_javacpp_cudnn_00024cudnnPoolingStruct_2Lorg_bytedeco_javacpp_cudnn_00024cudnnTensorStruct_2Ljava_nio_IntBuffer_2Ljava_nio_IntBuffer_2Ljava_nio_IntBuffer_2Ljava_nio_IntBuffer_2 in jnicudnn-910341.o
     (maybe you meant: _Java_org_bytedeco_javacpp_cudnn_cudnnGetPooling2dForwardOutputDim__Lorg_bytedeco_javacpp_cudnn_00024cudnnPoolingStruct_2Lorg_bytedeco_javacpp_cudnn_00024cudnnTensorStruct_2Ljava_nio_IntBuffer_2Ljava_nio_IntBuffer_2Ljava_nio_IntBuffer_2Ljava_nio_IntBuffer_2, _Java_org_bytedeco_javacpp_cudnn_cudnnGetPooling2dForwardOutputDim__Lorg_bytedeco_javacpp_cudnn_00024cudnnPoolingStruct_2Lorg_bytedeco_javacpp_cudnn_00024cudnnTensorStruct_2Lorg_bytedeco_javacpp_IntPointer_2Lorg_bytedeco_javacpp_IntPointer_2Lorg_bytedeco_javacpp_IntPointer_2Lorg_bytedeco_javacpp_IntPointer_2 , _Java_org_bytedeco_javacpp_cudnn_cudnnGetPooling2dForwardOutputDim__Lorg_bytedeco_javacpp_cudnn_00024cudnnPoolingStruct_2Lorg_bytedeco_javacpp_cudnn_00024cudnnTensorStruct_2_3I_3I_3I_3I )
  "_cudnnGetPoolingNdForwardOutputDim", referenced from:
      _Java_org_bytedeco_javacpp_cudnn_cudnnGetPoolingNdForwardOutputDim__Lorg_bytedeco_javacpp_cudnn_00024cudnnPoolingStruct_2Lorg_bytedeco_javacpp_cudnn_00024cudnnTensorStruct_2I_3I in jnicudnn-910341.o
      _Java_org_bytedeco_javacpp_cudnn_cudnnGetPoolingNdForwardOutputDim__Lorg_bytedeco_javacpp_cudnn_00024cudnnPoolingStruct_2Lorg_bytedeco_javacpp_cudnn_00024cudnnTensorStruct_2ILjava_nio_IntBuffer_2 in jnicudnn-910341.o
      _Java_org_bytedeco_javacpp_cudnn_cudnnGetPoolingNdForwardOutputDim__Lorg_bytedeco_javacpp_cudnn_00024cudnnPoolingStruct_2Lorg_bytedeco_javacpp_cudnn_00024cudnnTensorStruct_2ILorg_bytedeco_javacpp_IntPointer_2 in jnicudnn-910341.o
     (maybe you meant: _Java_org_bytedeco_javacpp_cudnn_cudnnGetPoolingNdForwardOutputDim__Lorg_bytedeco_javacpp_cudnn_00024cudnnPoolingStruct_2Lorg_bytedeco_javacpp_cudnn_00024cudnnTensorStruct_2ILjava_nio_IntBuffer_2, _Java_org_bytedeco_javacpp_cudnn_cudnnGetPoolingNdForwardOutputDim__Lorg_bytedeco_javacpp_cudnn_00024cudnnPoolingStruct_2Lorg_bytedeco_javacpp_cudnn_00024cudnnTensorStruct_2I_3I , _Java_org_bytedeco_javacpp_cudnn_cudnnGetPoolingNdForwardOutputDim__Lorg_bytedeco_javacpp_cudnn_00024cudnnPoolingStruct_2Lorg_bytedeco_javacpp_cudnn_00024cudnnTensorStruct_2ILorg_bytedeco_javacpp_IntPointer_2 )
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

```

**Comment by saudet** [comment]
Looks like there's a bug in cuDNN. Those symbols are not part of the shared libraries. Try to add this line in the `map()` method of the presets:
```java
    infoMap.put(new Info("cudnnGetPoolingNdForwardOutputDim", "cudnnGetPooling2dForwardOutputDim").skip());
```

**Comment by RoelVanderPaal** [comment]
cuddn compiles fine, but I get another error:
```
[INFO] Generating /Users/roel/workspace/javacpp-presets/cuda/target/classes/org/bytedeco/javacpp/jninppi.cpp
[INFO] Compiling /Users/roel/workspace/javacpp-presets/cuda/target/classes/org/bytedeco/javacpp/macosx-x86_64/libjninppi.dylib
[INFO] g++ -I/usr/local/cuda/include/ -I/Library/Java/JavaVirtualMachines/jdk1.8.0_20.jdk/Contents/Home/include -I/Library/Java/JavaVirtualMachines/jdk1.8.0_20.jdk/Contents/Home/include/darwin /Users/roel/workspace/javacpp-presets/cuda/target/classes/org/bytedeco/javacpp/jninppi.cpp -Wl,-rpath,@loader_path/. -march=x86-64 -m64 -Wall -O3 -fPIC -dynamiclib -o /Users/roel/workspace/javacpp-presets/cuda/target/classes/org/bytedeco/javacpp/macosx-x86_64/libjninppi.dylib -L/usr/local/cuda/lib/ -Wl,-rpath,/usr/local/cuda/lib/ -lnppi -lnppc -lcuda -lcudart -framework JavaVM 
Undefined symbols for architecture x86_64:
  "_nppiHLSToBGR_8u_AC4R", referenced from:
      _Java_org_bytedeco_javacpp_nppi_nppiHLSToBGR_18u_1AC4R___3BI_3BILorg_bytedeco_javacpp_nppc_00024NppiSize_2 in jninppi-52acd0.o
      _Java_org_bytedeco_javacpp_nppi_nppiHLSToBGR_18u_1AC4R__Ljava_nio_ByteBuffer_2ILjava_nio_ByteBuffer_2ILorg_bytedeco_javacpp_nppc_00024NppiSize_2 in jninppi-52acd0.o
      _Java_org_bytedeco_javacpp_nppi_nppiHLSToBGR_18u_1AC4R__Lorg_bytedeco_javacpp_BytePointer_2ILorg_bytedeco_javacpp_BytePointer_2ILorg_bytedeco_javacpp_nppc_00024NppiSize_2 in jninppi-52acd0.o
  "_nppiNormRelInfGetBufferHostSize_32s_C1R", referenced from:
      _Java_org_bytedeco_javacpp_nppi_nppiNormRelInfGetBufferHostSize_132s_1C1R__Lorg_bytedeco_javacpp_nppc_00024NppiSize_2_3I in jninppi-52acd0.o
      _Java_org_bytedeco_javacpp_nppi_nppiNormRelInfGetBufferHostSize_132s_1C1R__Lorg_bytedeco_javacpp_nppc_00024NppiSize_2Ljava_nio_IntBuffer_2 in jninppi-52acd0.o
      _Java_org_bytedeco_javacpp_nppi_nppiNormRelInfGetBufferHostSize_132s_1C1R__Lorg_bytedeco_javacpp_nppc_00024NppiSize_2Lorg_bytedeco_javacpp_IntPointer_2 in jninppi-52acd0.o
  "_nppiSub_32s_C4IRSfs", referenced from:
      _Java_org_bytedeco_javacpp_nppi_nppiSub_132s_1C4IRSfs___3II_3IILorg_bytedeco_javacpp_nppc_00024NppiSize_2I in jninppi-52acd0.o
      _Java_org_bytedeco_javacpp_nppi_nppiSub_132s_1C4IRSfs__Lorg_bytedeco_javacpp_IntPointer_2ILorg_bytedeco_javacpp_IntPointer_2ILorg_bytedeco_javacpp_nppc_00024NppiSize_2I in jninppi-52acd0.o
      _Java_org_bytedeco_javacpp_nppi_nppiSub_132s_1C4IRSfs__Ljava_nio_IntBuffer_2ILjava_nio_IntBuffer_2ILorg_bytedeco_javacpp_nppc_00024NppiSize_2I in jninppi-52acd0.o
  "_nppiSub_32s_C4RSfs", referenced from:
      _Java_org_bytedeco_javacpp_nppi_nppiSub_132s_1C4RSfs___3II_3II_3IILorg_bytedeco_javacpp_nppc_00024NppiSize_2I in jninppi-52acd0.o
      _Java_org_bytedeco_javacpp_nppi_nppiSub_132s_1C4RSfs__Ljava_nio_IntBuffer_2ILjava_nio_IntBuffer_2ILjava_nio_IntBuffer_2ILorg_bytedeco_javacpp_nppc_00024NppiSize_2I in jninppi-52acd0.o
      _Java_org_bytedeco_javacpp_nppi_nppiSub_132s_1C4RSfs__Lorg_bytedeco_javacpp_IntPointer_2ILorg_bytedeco_javacpp_IntPointer_2ILorg_bytedeco_javacpp_IntPointer_2ILorg_bytedeco_javacpp_nppc_00024NppiSize_2I in jninppi-52acd0.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by RoelVanderPaal** [comment]
I added this line to nppi.java and now it compiles fine:
```
infoMap.put(new Info("nppiHLSToBGR_8u_AC4R", "nppiNormRelInfGetBufferHostSize_32s_C1R", "nppiSub_32s_C4IRSfs", "nppiSub_32s_C4RSfs").skip());
```

**Comment by saudet** [comment]
Yeah, looks like there's missing exports for NPP as well. Thanks for testing this out :) Linux doesn't throw an error or even a warning on missing symbols like that.

**Comment by saudet** [comment]
Fix included in version 1.0. Thanks for reporting and enjoy!

**Comment by cnuernber** [comment]
This isn't a great fix; perhaps cudnn 5 includes these in the new system?  Double checking that the expected sizes coming out of the convolution and pooling layers avoids memory overwrite errors and other sort of nasty issues.
